feat(Anomaly-Detection): add Online Isolation Forest

ineveLoppiliF opened the pull request.

ineveLoppiliF commented:
Hi everyone,
It looks like there is an issue with a library used in the Online Isolation Forest implementation ("ModuleNotFoundError: No module named 'tqdm'"). Should I remove the dependencies on this library and open a new Pull Request?
Thank you.

justinuliu commented:
> Hi everyone, It looks like there is an issue with a library used in the Online Isolation Forest implementation ("ModuleNotFoundError: No module named 'tqdm'"). Should I remove the dependencies on this library and open a new Pull Request? Thank you.

Hi Filippo, If the tqdm library is **not** needed to implement the Online Isolation Forest, I think we should remove it.

hmgomes commented:
Hi @ineveLoppiliF, thanks once again for the contribution, I’ve completed reviewing it. 
We need to make some adjustments to the API to make sure it is aligned with other learners. Also, we might be able to simplify the code and concentrate it in a single file. See below some suggestions: 

1. Renaming it to OnlineIsolationForest would make it easier for those familiar with IsolationForest to recognize the algorithm.
2. All code should be consolidated into one file, e.g., _online_isolation_forest.py.
3. Abstract methods from OnlineIsolationForest should be added to AnomalyDetector, with code from BoundedRandomProjectionOnlineIForest moved there.
4. I noticed you've structured it generically, but currently, only one algorithm (BoundedRandomProjection) is implemented. * 5. Are we anticipating additional implementations in the future?
6. Should BoundedRandomProjectionOnlineIForest be made the default implementation of OnlineIsolationForest?
7. It would be preferable to avoid using static create methods to maintain consistency with other algorithms. This way, creating an OnlineIsolationForest could be as straightforward as OnlineIsolationForest(...), rather than OnlineIsolationForest.create(...).

I have implemented these suggestions already. If you want I can make a commit with them (number 3 required changes to the base method and to HSTrees)

ineveLoppiliF commented:
Hi @hmgomes, thank you for your review and for simplifying the code!

I would like to clarify that I structured the code generically because, depending on the split criteria chosen, it is possible to "isolate" instances based on different distance metrics. While the split criteria used by IsolationForest (and thus by OnlineIsolationForest) is closely tied to the L1-norm, instances can, in general, be considered as "isolated" with respect to other distances, such as other L-norms, Hamming, Cosine, etc.
Apart from that, I confirm that the version of OnlineIsolationForest presented in the recently accepted paper is the one called BoundedRandomProjectionOnlineIForest. The same applies for OnlineIsolationTree (BoundedRandomProjectionOnlineITree) and OnlineIsolationNode (BoundedRandomProjectionOnlineINode).

If our current focus is to integrate OnlineIsolationForest into the library, and avoid confusing external readers who are not aware of potential future extensions, I suggest removing the generic parts from OnlineIsolationTree and OnlineIsolationNode as well.
If you agree, I can remove the generic parts from the OnlineIsolation* components, and commit the updates.

Thank you again!

hmgomes commented:
Hi @ineveLoppiliF 

Thanks for the clarifications. Regarding the following:

> If our current focus is to integrate OnlineIsolationForest into the library, and avoid confusing external readers who are not aware of potential future extensions, I suggest removing the generic parts from OnlineIsolationTree and OnlineIsolationNode as well.
> If you agree, I can remove the generic parts from the OnlineIsolation* components, and commit the updates.

Please proceed to streamline the generic parts and commit the updates

Thanks again for your contribution

Heitor

tachyonicClock commented on src/capymoa/anomaly/online_isolation_forest.py:
Could this file be renamed to `_online_isolation_forest.py`. This avoid duplicate import paths i.e:
```
from capymoa.anomaly import OnlineIsolationForest
from capymoa.anomaly.online_isolation_forest import OnlineIsolationForest
```

tachyonicClock replied:
Some of your functions should be "private" if they should not be called by a user i.e prepend with `_` `_my_func`

tachyonicClock commented on src/capymoa/anomaly/online_isolation_forest.py:
```diff
@@ -0,0 +1,341 @@
+    AUC: 0.52
+    """
+    def __init__(self, schema: Schema = None, random_seed: int = 1, num_trees: int = 32, max_leaf_samples: int = 32,
+                 type: str = 'adaptive', subsample: float = 1.0, window_size: int = 2048, branching_factor: int = 2,
```
- `type` and `metric` should be a `typing.Literal` e.g `Literal["fixed", "adaptive]` to document the options. 
- The variable name `type` should be avoided because it overrides a built in function. 
- `metric` is not documented in the docstring

tachyonicClock commented on src/capymoa/anomaly/online_isolation_forest.py:
```diff
@@ -0,0 +1,341 @@
+    an ensemble anomaly detector capable of adapting to concept drift.
+
+    Reference:
+    @inproceedings{Leveni2024,
```
A format like APA would be easier to read in generate documentation than bibtex

tachyonicClock commented on src/capymoa/anomaly/online_isolation_forest.py:
```diff
@@ -0,0 +1,341 @@
+                node_index, node.children[i] = self.recursive_learn(node.children[i], data[indices], node_index)
+            return node_index, node
+
+    def recursive_build(self, data: ndarray, depth: int = 0, node_index: int = 0) -> (int, OnlineIsolationNode):
```
My IDE complains that ``def foo() -> (int, str)`` should be  ``def foo() -> Tuple[int, str] :``. `Tuple` is from `typing`

tachyonicClock commented on src/capymoa/anomaly/online_isolation_forest.py:
```diff
@@ -0,0 +1,341 @@
+        return partition
+
+
+class OnlineIsolationNode:
```
Consider using a data-class here https://docs.python.org/3/library/dataclasses.html.

tachyonicClock replied:
Many of these types should be `Optional[]` since they can be `None`. The issue is that this makes following the logic harder because the types are wrong. e.g `if node.children is None:` should always be `False`

tachyonicClock commented on src/capymoa/anomaly/online_isolation_forest.py:
```diff
@@ -0,0 +1,341 @@
+    year         = {2024},
+    organization = {Proceedings of Machine Learning Research (PMLR)}}
+    Example usage:
+    >>> from capymoa.datasets import ElectricityTiny
```
Examples are awesome thank you.

tachyonicClock commented on src/capymoa/anomaly/online_isolation_forest.py:
```diff
@@ -0,0 +1,341 @@
+
+    def score_instance(self, instance: Instance) -> AnomalyScore:
+        data: ndarray = instance.x.reshape((1, -1))
+        return self.score_batch(data)
```
This should convert this to a double/float before returning it. The implicit conversion will be an error in a future version of numpy
```
Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future.
```

tachyonicClock reviewed (changes requested):
Thanks for taking the time to implement this for us. It is good work. I've made some minor suggestions.

tachyonicClock commented on tests/test_anomaly_detectors.py:
```diff
@@ -17,9 +18,11 @@
     "learner_constructor,auc,cli_string",
     [
         (partial(HalfSpaceTrees, window_size=100, number_of_trees=25, max_depth=15), 0.54, None),
+        (partial(OnlineIsolationForest, num_trees=32, max_leaf_samples=32, window_size=2048), 0.52, None),
```
This window size is maybe too large the dataset ``ElectricityTiny`` is only 1k long.

tachyonicClock commented:
I tested the parallel feature and it seems to be unhelpful. Could this be a bi-product of the test case being a very small dataset? Is this what you would expect?

```python
@pytest.mark.parametrize(
    "learner_constructor,auc,cli_string",
    [
        (partial(HalfSpaceTrees, window_size=100, number_of_trees=25, max_depth=15), 0.54, None),
        (partial(OnlineIsolationForest, num_trees=32, max_leaf_samples=32, window_size=2048), 0.52, None),
        (partial(OnlineIsolationForest, num_trees=32, max_leaf_samples=32, window_size=2048, n_jobs=-1), 0.52, None),
    ],
    ids=[
        "HalfSpaceTrees",
        "OnlineIsolationForest",
        "OnlineIsolationForest(n_jobs=-1)",
    ],
)
....
```

```console
$pytest -k OnlineIsolationForest --durations=0
...
44.54s call     tests/test_anomaly_detectors.py::test_anomaly_detectors[OnlineIsolationForest(n_jobs=-1)]
22.25s call     tests/test_anomaly_detectors.py::test_anomaly_detectors[OnlineIsolationForest]
...
```

hmgomes commented:
Hi @ineveLoppiliF , I asked @tachyonicClock to have a second look at the code and here is the results of his review. These changes should be minor, so if you want we can do these (i.e. merge this PR and then create another one for the minor changes) or if you prefer to do it, that is also fine. 

Thanks again! 
Heitor

ineveLoppiliF commented:
Hi @tachyonicClock, thank you very much for the review. Your advices has been very helpful, and I think I have implemented all of them.

ineveLoppiliF commented:
Hi @tachyonicClock, thank you very much for the review. Your advice has been very helpful, and I believe I have implemented all of your suggestions.

Regarding the parallel feature, I think you're right. I implemented it because it seemed reasonable to me to allow for the parallel construction of OnlineIsolationTrees, but I haven't found it useful in practice. My guess is that the NumPy functions underlying OnlineIsolationTree already parallelize execution across multiple processors, leading to overhead when reallocating processors for different OnlineIsolationTree constructions, which nullifies the parallel gains. However, I haven't investigated this thoroughly, so I might be wrong.

Thank you!
Filippo Leveni

tachyonicClock commented:
Awesome. Thanks for taking the time to make those changes. I went ahead and merged it because it looks good and I don't think it impacts anyone else's work (@hmgomes).
